app/contacts/contact_slack.py

```diff
@@ -0,0 +1,302 @@
+
+    async def _post_instructions(self, text, paw):
+        try:
+            if await self._wait_for_paw(paw, comm_type='instructions'):
+                return
+            s = await self._post_slack_message(self._build_slack_message(comm_type='instructions', paw=paw,
+                                                                         data=text))
+            return s
+        except Exception as e:
+            self.log.warning('Posting instructions over c2 (%s) failed!: %s' % (self.__class__.__name__, e))
+
+    async def _send_payloads(self, agent, instructions):
+        for i in instructions:
+            for p in i.payloads:
+                filename, payload_contents = await self._get_payload_content(p, agent)
+                await self._post_payloads(filename, payload_contents, '%s-%s' % (agent.paw, filename))
+
+    async def _post_payloads(self, filename, payload_contents, paw):
+        try:
+            if await self._wait_for_paw(paw, comm_type='payloads'):
+                return
+            s = await self._post_slack(self._build_slack_content(comm_type='payloads', paw=paw, files=self._encode_string(payload_contents)))
+            return s
+        except Exception as e:
+            self.log.warning('Posting payload over c2 (%s) failed! %s' % (self.__class__.__name__, e))
+
+    async def _store_file_chunk(self, paw, upload_id, filename, contents, curr_chunk, total_chunks):
+        pending_upload = self.pending_uploads[paw].get(upload_id)
+        if not pending_upload:
+            # starting brand new upload
+            pending_upload = self.SlackUpload(upload_id, filename, total_chunks)
+            self.pending_uploads[paw][upload_id] = pending_upload
+        pending_upload.add_chunk(curr_chunk - 1, contents)
+
+    async def _ready_to_export(self, paw, upload_id):
+        pending_upload = self.pending_uploads[paw].get(upload_id)
+        return pending_upload is not None and pending_upload.is_complete() and not pending_upload.exported
+
+    async def _submit_uploaded_file(self, paw, upload_id):
+        upload_info = self.pending_uploads[paw].get(upload_id)
+        if upload_info is not None:
+            created_dir = os.path.normpath('/' + paw).lstrip('/')
+            saveto_dir = await self.file_svc.create_exfil_sub_directory(dir_name=created_dir)
+            unique_filename = ''.join([upload_info.filename, '-', upload_id[0:10]])
+            await self.file_svc.save_file(unique_filename, upload_info.export_contents(), saveto_dir)
+            self.log.debug('Uploaded file %s/%s' % (saveto_dir, upload_info.filename))
+
+    async def _wait_for_paw(self, paw, comm_type):
+        for message in await self._get_slack():
+            if '{}-{}'.format(comm_type, paw) == message['text'].split(' | ')[0]:
+                return True
+        return False
+
+    async def _get_slack_data(self, comm_type):
+        data = await self._get_raw_slack_data(comm_type=comm_type)
```

Comment:
these two lines could be their own separate function `_get_and_delete_raw_slack_data`, since they're also repeated in the same order under `_get_slack_content`